File: src/main/java/com/bajdcc/LALR1/grammar/runtime/RuntimeCodePage.java
package com.bajdcc.LALR1.grammar.runtime;

import com.bajdcc.util.intervalTree.Interval;
import com.bajdcc.util.intervalTree.IntervalTree;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;
import java.util.List;
import java.util.stream.Collectors;

/**
 * 【目标代码】代码页
 *
 * @author bajdcc
 */
public class RuntimeCodePage implements Serializable {

	private static final long serialVersionUID = 1L;

	/**
	 * 数据
	 */
	private List<Object> data;

	/**
	 * 指令
	 */
	private List<Byte> insts;

	/**
	 * 调试开发
	 */
	private IRuntimeDebugInfo info;

	/**
	 * 汇编调试的符号表
	 */
	private List<Interval<Object>> itvList;

	private transient IntervalTree<Object> tree = null;

	public RuntimeCodePage(List<Object> data, List<Byte> insts,
	                       IRuntimeDebugInfo info, List<Interval<Object>> itvList) {
		this.data = data;
		this.insts = insts;
		this.info = info;
		this.itvList = itvList;
	}

	public static RuntimeCodePage importFromStream(InputStream input) {
		return ObjectTools.deserialize(input);
	}

	public static void exportFromStream(RuntimeCodePage page,
	                                    OutputStream output) {
		ObjectTools.serialize(page, output);
	}

	public List<Object> getData() {
		return data;
	}

	public List<Byte> getInsts() {
		return insts;
	}

	public String getDebugInfoByInc(long index) {
		if (tree == null)
			tree = new IntervalTree<>(itvList);
		List<Object> list = tree.get(index);
		if (list.isEmpty())
			return "[NO DEBUG INFO]";
		return list.stream().map(String::valueOf).map(a -> "---=== 代码片段 ===---" + (System.lineSeparator() + a)).collect(Collectors.joining(System.lineSeparator()));
	}

	public IRuntimeDebugInfo getInfo() {
		return info;
	}

	public String getCodeString() {
		return "#### 目标代码 ####" +
				System.lineSeparator() +
				insts +
				System.lineSeparator();
	}

	@Override
	public String toString() {
		return getCodeString();
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/runtime/RuntimeCodePage.kt
package com.bajdcc.LALR1.grammar.runtime

import com.bajdcc.util.intervalTree.Interval
import com.bajdcc.util.intervalTree.IntervalTree
import java.io.InputStream
import java.io.OutputStream
import java.io.Serializable

/**
 * 【目标代码】代码页
 *
 * @author bajdcc
 */
class RuntimeCodePage(
        /**
         * 数据
         */
        val data: List<Any>,
        /**
         * 指令
         */
        val insts: List<Byte>,
        /**
         * 调试开发
         */
        val info: IRuntimeDebugInfo,
        /**
         * 汇编调试的符号表
         */
        private val itvList: List<Interval<Any>>) : Serializable {

    @Transient
    private var tree: IntervalTree<Any>? = null

    val codeString: String
        get() = "#### 目标代码 ####" +
                System.lineSeparator() +
                insts +
                System.lineSeparator()

    fun getDebugInfoByInc(index: Long): String {
        if (tree == null)
            tree = IntervalTree(itvList)
        val list = tree!![index]
        return if (list.isEmpty()) "[NO DEBUG INFO]" else
            list.joinToString(
                    separator = System.lineSeparator(),
                    prefix = "---=== 代码片段 ===---" + System.lineSeparator(),
                    transform = { it.toString() })
    }

    override fun toString(): String {
        return codeString
    }

    companion object {

        private const val serialVersionUID = 1L

        fun importFromStream(input: InputStream): RuntimeCodePage {
            return ObjectTools.deserialize<RuntimeCodePage>(input)!!
        }

        fun exportFromStream(page: RuntimeCodePage,
                             output: OutputStream) {
            ObjectTools.serialize(page, output)
        }
    }
}
